lib/storage/src/content_manager/consensus_manager.rs

```diff
@@ -586,6 +586,44 @@ impl<C: CollectionContainer> ConsensusManager<C> {
         }
     }
 
+    /// Wait and block until consensus reaches a `commit` and `term`
+    ///
+    /// # Returns
+    ///
+    /// Returns `true` if successful.
+    /// Returns `false` on failure, if we have diverged commit/term for example.
+    pub async fn wait_for_consensus_commit(
+        &self,
+        commit: u64,
+        term: u64,
+        consensus_tick: Duration,
+        timeout: Duration,
+    ) -> bool {
+        let start = Instant::now();
+
+        // TODO: naive approach with spinlock for waiting on commit/term, find better way
+        while start.elapsed() < timeout {
+            let state = &self.hard_state();
+
+            // Okay if we're on a newer term, or if commit is reached within same term
+            let is_ok = state.term > term || (state.term == term && state.commit >= commit);
+            if is_ok {
+                return true;
+            }
+
+            // Fail if commit is reached but we're on an older term
+            let is_fail = state.term < term && state.commit >= commit;
+            if is_fail {
+                return false;
+            }
```

Comment:
I think these checks are not fully correct.

- AFAIK, `state.term > term` does not mean that commit is reached.
  - The node may have lost connectivity with the leader and started an election.
  - When it does so, _it increments its own term_ and sends `RequestVote` requests to other nodes.
  - (And after that, the other nodes would deny this node a vote, because it's `commit` is lagging behind.)
- Same with the `state.term < term && state.commit >= commit`.
  - The commit _is_ reached. It won't be rolled back.
    - (IIRC, it is possible under some corner cases, but generally, what has been committed stays committed.)
  - And the term would catch up soon.
    - (Literally, when the next message from the current node propagate to this node.)

<details>
<summary>Images</summary>

Visualization from raft.github.io. Node `S2` incremented the term and started an election.
<img width="818" alt="Screenshot 2023-08-24 at 13 23 42" src="https://github.com/qdrant/qdrant/assets/2725918/aff59cec-7638-49e8-b86f-e05f4a8bfd4e">

Image from the Raft paper. Notice the very last point under "Receiver implementation".
<img width="394" alt="Screenshot 2023-08-24 at 13 23 21" src="https://github.com/qdrant/qdrant/assets/2725918/d390a2ef-506a-4785-89fe-492a22c734ee">

</details>

Reply:
Thanks for diving into this.

I agree that this isn't perfect, but we have limited information to work with. Do you have a suggestion on how to improve this 'waiting' process?

> AFAIK, state.term > term does not mean that commit is reached. 

Correct. The idea was that it does mean that it is at least more up-to-date.

I'm not sure how to improve on this. Maybe we'll have to include the current selected leader, so both parties must agree on that as well.

> Same with the state.term < term && state.commit >= commit. 

Isn't it possible for the node we wait upon to fail to promote to the latest term? In that case the current (higher) commit ID is meaningless.

By the way, the overall logic here is sourced from https://github.com/tikv/raft-rs/blob/cd621e4c7bdf4278e970e13ea236bb0559c9f21b/src/raft_log.rs#L412-L420

Anyway. I may very well be missing something important here. I'd love some input on how we can make this call work!

Reply:
> The idea was that it does mean that it is at least more up-to-date.

Imagine:

- There are 3 nodes
  - node 1 is the leader
  - node 2 is up-to-date follower
  - node 3 is lagging-behind follower
- There's a network issue and node 3 lost _Raft_ connectivity
- Node 3 does not receive heartbeats from leader for a while, increases its `term` and start election
- We send `wait_for_consensus_commit` to node 3
- Is node 3 more up-to-date than nodes 1 and 2?
  - I don't think so.

Reply:
My point is that `term` increments independently of `commit`, and it can happen on _any_ node at _any_ time. It's meaningless to consider some node more-up-to date than some other node based _only_ on `term`.

Reply:
> > The idea was that it does mean that it is at least more up-to-date.
> 
> Imagine:
> 
>     * There are 3 nodes
>       
>       * node 1 is the leader
>       * node 2 is up-to-date follower
>       * node 3 is lagging-behind follower
> 
>     * There's a network issue and node 3 lost _Raft_ connectivity
> 
>     * Node 3 does not receive heartbeats from leader for a while, increases its `term` and start election
> 
>     * We send `wait_for_consensus_commit` to node 3
> 
>     * Is node 3 more up-to-date than nodes 1 and 2?
>       
>       * I don't think so.

Good example. I believe that including the current leader as part of the comparison would mitigate this. Can we think of other examples that would still be problematic with that additional check?

Reply:
> Isn't it possible for the node we wait upon to fail to promote to the latest term?

I'm not _100%_ sure, but AFAIK, _any_ node should increase its term when it receives _any_ message with higher term.

> In that case the current (higher) commit ID is meaningless.

What about above example with 3 nodes, where node 3 have higher `term`, but lower `commit? Is node 1 or 2 behind node 3? Node 3 won't be promoted to leader, so node 3 having higher `term` is a transient state.

Reply:
what's the recommendation here? Can we use commit + term to ensure the state equivalency? As I understood, from @ffuugoo explanation - term can be misleading, but commit rolls back never enough

Reply:
> Can we use commit + term to ensure the state equivalency?

I think it would work better, but I'll only be able to tell for sure if I spend a day or so working through all possible scenarios and reading Raft paper. :/

But so far `term + commit` check seems more solid, than just the `term` check.

Reply:
Yep, that what I was thinking about.

Reply:
Implemented in 2074c48f085136075fb07472c0207fae83b52cae